### PR TITLE
trailing-whitespace hook: restoring original file in case of failure - fixes #134

### LINKED ISSUE #134: Non-utf8 file will cause trailing-whitespace to truncate the file contents
Hi,

(I started investigating this in https://github.com/pre-commit/pre-commit/issues/397)

Bug reproductions steps, after setting up a git repository with the "trailing-whitespace" pre-commit hook installed (with the latest sha: 775a7906cd6004bec2401a85ed16d5b3540d2f94):
```
echo -en '\x3c\x61\x3e\xe9\x20\x0a\x3c\x2f\x61\x3e' > minimal_crasher.xml
git add minimal_crasher.xml
pre-commit run trailing-whitespace --files minimal_crasher.xml
```

Results in:
````
Trim Trailing Whitespace...............................................................................................................................................................Failed
hookid: trailing-whitespace

Files were modified by this hook. Additional output:

Fixing minimal_crasher.xml
Traceback (most recent call last):
  File "/home/lucas_cimon/.pre-commit/repok3wjweh4/py_env-default/bin/trailing-whitespace-fixer", line 9, in <module>
    load_entry_point('pre-commit-hooks==0.6.0', 'console_scripts', 'trailing-whitespace-fixer')()
  File "/home/lucas_cimon/.pre-commit/repok3wjweh4/py_env-default/lib/python3.4/site-packages/pre_commit_hooks/trailing_whitespace_fixer.py", line 71, in fix_trailing_whitespace
    _fix_file(bad_whitespace_file, all_markdown or extension in md_exts)
  File "/home/lucas_cimon/.pre-commit/repok3wjweh4/py_env-default/lib/python3.4/site-packages/pre_commit_hooks/trailing_whitespace_fixer.py", line 12, in _fix_file
    for line in fileinput.input([filename], inplace=True):
  File "/usr/lib/python3.4/fileinput.py", line 263, in __next__
    line = self.readline()
  File "/usr/lib/python3.4/fileinput.py", line 363, in readline
    self._buffer = self._file.readlines(self._bufsize)
  File "/home/lucas_cimon/.pre-commit/repok3wjweh4/py_env-default/lib/python3.4/codecs.py", line 319, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 3: invalid continuation byte
```

And **all the content of `minimal_crasher.xml` lost** !!

In this test the xml file is very small, but it can happen with any file size nor number of staged modifications !

### PR DESCRIPTION
**Pull request by Lucas-C**

### CODE REVIEW

**Comment by asottile** [comment]
Needs a test, also I'm not the biggest fan of fileinput, I think it might be easier just using `io.open` instead?  (that'd remove the awkward temporary file that may clobber other files in the working directory).

**Comment by Lucas-C** [comment]
Test added

**Comment by Lucas-C** [comment]
I'll do the move to `io.open` in a 2nd commit, when implementing the "option 2" we discussed.

**Comment by asottile** [comment]
New test seems falling :(

**Comment by Lucas-C** [comment]
Only the py34 AppVeyor test remains to fail.
I don't really get why actually

**Comment by asottile** [comment]
When I'm looking for weird bugs that's music to my ears. 

I believe it's due to windows having a default encoding of cp1252

**Comment by asottile** [comment]
Maybe it's time to approach the `'rb'` approach ;)

**Comment by Lucas-C** [comment]
Done :)

**Comment by Lucas-C** [comment]
AppVeyor is still failing :(
I wonder if it's because `readlines()` mangle newline bytes under Windows (it does not under Cygwin).
Sadly I'm going to be away from any computer for a week, so I won't make any progress on this in the following days.

**Comment by asottile** [comment]
Ah yeah I've hit this before, the `newline=''` kwarg to open might fix this

**Comment by asottile** [comment]
I'm off holiday in a few days so maybe I'll try and pick it up then

File: pre_commit_hooks/trailing_whitespace_fixer.py [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
-        # preserve trailing two-space for non-blank lines in markdown files
-        if markdown and (not line.isspace()) and (line.endswith("  \n")):
-            line = line.rstrip(' \n')
-            # only preserve if there are no trailing tabs or unusual whitespace
-            if not line[-1].isspace():
-                print(line + "  ")
-                continue
-
-        print(line.rstrip())
+    with tempfile.NamedTemporaryFile(mode='w+b', delete=False) as tmp_file:
```
Comment (asottile) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: just store the old contents in memory imo

File: pre_commit_hooks/trailing_whitespace_fixer.py [hunk 2]
```diff
@@ -1,24 +1,28 @@
...
+        with open(filename, mode='rb') as original_file:
+            for line in original_file.readlines():
+                # preserve trailing two-space for non-blank lines in markdown files
+                eol = b'\r\n' if line[-2:] == b'\r\n' else b'\n'
+                if markdown and (not line.isspace()) and line.endswith(b'  ' + eol):
+                    line = line.rstrip(b' \r\n')  # restricted stripping: e.g. \t are not stripped
```
Comment (asottile) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: weird, I realize the lhs does this but I cannot remember why, should we also strip tabs?

File: pre_commit_hooks/trailing_whitespace_fixer.py (hunk 1 above)
Comment (Lucas-C) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: Ok

File: pre_commit_hooks/trailing_whitespace_fixer.py (hunk 2 above)
Comment (Lucas-C) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: If we replace this by a simple `rstrip()`, the following test fail:
```
    def test_markdown_linebreak_ext_opt(filename='bar.MD', input_s='bar   \nbaz\t   \n\t\n', output='bar  \nbaz\n\n')):
        ...
>       assert path.read() == output
E       assert 'bar  \nbaz  \n\n' == 'bar  \nbaz\n\n'
E           bar
E         - baz
E         ?    --
E         + baz
```

Comment (asottile) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: either way, I wouldn't worry about this if it means changing a test.  We can address that in a followup or not at all :)

Comment (Lucas-C) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: OK, fixes pushed

File: pre_commit_hooks/trailing_whitespace_fixer.py [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
+        for line in lines:
+            file_processed.write(line)
 
+def _process_line(line, is_markdown):
+    # preserve trailing two-space for non-blank lines in markdown files
+    eol = b'\r\n' if line[-2:] == b'\r\n' else b'\n'
+    if is_markdown and (not line.isspace()) and line.endswith(b'  ' + eol):
+        line = line.rstrip()
+        # only preserve if there are no trailing tabs or unusual whitespace
+        if not line[-1:].isspace():
```
Comment (asottile) [inline pre_commit_hooks/trailing_whitespace_fixer.py]: seems (from coverage) this is always true

**Comment by asottile** [comment]
Stellar!  Thanks for the hard work on this :)

**Pull request merged**